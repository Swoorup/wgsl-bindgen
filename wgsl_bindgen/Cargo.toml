[package]
name = "wgsl_bindgen"
authors.workspace = true
version.workspace = true
edition.workspace = true
repository.workspace = true
documentation.workspace = true
license.workspace = true
description = "Type safe Rust bindings workflow for wgsl shaders in wgpu"
readme = "../README.md"

[dependencies]
naga = { workspace = true, features = ["wgsl-in"] }
wgpu-types.workspace = true
syn.workspace = true
quote.workspace = true
proc-macro2.workspace = true
prettyplease.workspace = true
thiserror.workspace = true
case.workspace = true
naga_oil.workspace = true
regex.workspace = true
data-encoding.workspace = true
indexmap.workspace = true
smallvec.workspace = true
derive_more.workspace = true
fxhash.workspace = true
derivative.workspace = true
smol_str.workspace = true
colored.workspace = true
derive_builder.workspace = true
miette.workspace = true
blake3.workspace = true
regex-syntax.workspace = true
strum.workspace = true
strum_macros.workspace = true
pathdiff.workspace = true
enumflags2.workspace = true
heck.workspace = true

[dev-dependencies]
indoc.workspace = true
pretty_assertions.workspace = true
miette = { workspace = true, features = ["fancy", "syntect"] }
