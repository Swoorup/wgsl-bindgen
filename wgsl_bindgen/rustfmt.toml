indent_style = "Block"
reorder_imports = true
tab_spaces = 2
max_width=90
fn_call_width=80
group_imports="StdExternalCrate"
