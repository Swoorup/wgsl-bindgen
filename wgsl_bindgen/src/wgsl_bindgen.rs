use std::io::Write;
use std::path::PathBuf;

use derive_builder::Builder;
use miette::Diagnostic;
use naga_oil::compose::{
  ComposableModuleDescriptor, Composer, ComposerError, NagaModuleDescriptor,
  ShaderLanguage,
};
use thiserror::Error;

use self::source_file::SourceFile;
use crate::{
  bevy_util::*, WgslEntryResult, WgslTypeMap, WgslTypeMapBuild, WgslTypeSerializeStrategy,
};
use crate::{create_rust_bindings, CreateModuleError, SourceFilePath};

const PKG_VER: &str = env!("CARGO_PKG_VERSION");
const PKG_NAME: &str = env!("CARGO_PKG_NAME");

/// Enum representing the possible errors that can occur in the `wgsl_bindgen` process.
///
/// This enum is used to represent all the different kinds of errors that can occur
/// when parsing WGSL shaders, generating Rust bindings, or performing other operations
/// in `wgsl_bindgen`.
#[derive(Debug, Error, Diagnostic)]
pub enum WgslBindgenError {
  #[error("All required fields need to be set upfront: {0}")]
  OptionBuilderError(#[from] WgslBindgenOptionBuilderError),

  #[error(transparent)]
  #[diagnostic(transparent)]
  DependencyTreeError(#[from] DependencyTreeError),

  #[error("Failed to compose modules with entry `{entry}`\n{msg}")]
  NagaModuleComposeError {
    entry: String,
    msg: String,
    inner: naga_oil::compose::ComposerErrorInner,
  },

  #[error(transparent)]
  ModuleCreationError(#[from] CreateModuleError),

  #[error(transparent)]
  WriteOutputError(#[from] std::io::Error),

  #[error("Output file is not specified. Maybe use `generate_string` instead")]
  OutputFileNotSpecified,
}

#[derive(Default, Debug, Clone, Copy, PartialEq, Eq)]
pub enum WgslShaderSourceOutputType {
  /// Include the final shader string directly in the output
  #[default]
  FinalShaderString,

  /// Use Composer including helper functions which will be executed on runtime
  Composer,
}

/// A struct representing a directory to scan for additional source files.
///
/// This struct is used to represent a directory to scan for additional source files
/// when generating Rust bindings for WGSL shaders. The `module_import_root` field
/// is used to specify the root prefix or namespace that should be applied to all
/// shaders given as the entrypoints, and the `directory` field is used to specify
/// the directory to scan for additional source files.
#[derive(Debug, Clone, Default)]
pub struct AdditionalScanDirectory {
  pub module_import_root: Option<String>,
  pub directory: String,
}

impl From<(Option<&str>, &str)> for AdditionalScanDirectory {
  fn from((module_import_root, directory): (Option<&str>, &str)) -> Self {
    Self {
      module_import_root: module_import_root.map(ToString::to_string),
      directory: directory.to_string(),
    }
  }
}

#[derive(Debug, Default, Builder)]
#[builder(
  setter(into),
  field(private),
  build_fn(private, name = "fallible_build")
)]
pub struct WgslBindgenOption {
  /// A vector of entry points to be added. Each entry point is represented as a `String`.
  #[builder(setter(each(name = "add_entry_point", into)))]
  pub entry_points: Vec<String>,

  /// The root prefix/namespace if any applied to all shaders given as the entrypoints.
  #[builder(default, setter(strip_option, into))]
  pub module_import_root: Option<String>,

  /// A boolean flag indicating whether to emit a rerun-if-changed directive to Cargo. Defaults to `true`.
  #[builder(default = "true")]
  pub emit_rerun_if_change: bool,

  /// A boolean flag indicating whether to skip header comments. Enabling headers allows to not rerun if contents did not change.
  #[builder(default = "false")]
  pub skip_header_comments: bool,

  /// A boolean flag indicating whether to skip the hash check. This will avoid reruns of bindings generation if
  /// entry shaders including their imports has not changed. Defaults to `false`.
  #[builder(default = "false")]
  pub skip_hash_check: bool,

  /// Derive [encase::ShaderType](https://docs.rs/encase/latest/encase/trait.ShaderType.html#)
  /// for user defined WGSL structs when `WgslTypeSerializeStrategy::Encase`.
  /// else derive bytemuck
  #[builder(default)]
  pub serialization_strategy: WgslTypeSerializeStrategy,

  /// Derive [serde::Serialize](https://docs.rs/serde/1.0.159/serde/trait.Serialize.html)
  /// and [serde::Deserialize](https://docs.rs/serde/1.0.159/serde/trait.Deserialize.html)
  /// for user defined WGSL structs when `true`.
  #[builder(default = "false")]
  pub derive_serde: bool,

  /// The type of output for the shader source. Defaults to `FinalShaderString`.
  #[builder(default)]
  pub shader_source_output_type: WgslShaderSourceOutputType,

  /// A mapping operation for WGSL built-in types. This is used to map WGSL built-in types to their corresponding representations.
  #[builder(setter(custom))]
  pub wgsl_type_map: WgslTypeMap,

  /// The output file path for the generated Rust bindings. Defaults to `None`.
  #[builder(default, setter(strip_option, into))]
  pub output_file: Option<PathBuf>,

  /// The additional set of directories to scan for source files.
  #[builder(default, setter(into, each(name = "additional_scan_dir", into)))]
  pub additional_scan_dirs: Vec<AdditionalScanDirectory>,
}

impl WgslBindgenOptionBuilder {
  pub fn build(&self) -> Result<WGSLBindgen, WgslBindgenError> {
    let options = self.fallible_build()?;
    WGSLBindgen::new(options)
  }

  pub fn wgsl_type_map(&mut self, map_build: impl WgslTypeMapBuild) -> &mut Self {
    let serialization_strategy = self
      .serialization_strategy
      .expect("Serialization strategy must be set before `wgs_type_map`");

    self.wgsl_type_map = Some(map_build.build(serialization_strategy));
    self
  }
}

pub struct WGSLBindgen {
  dependency_tree: DependencyTree,
  options: WgslBindgenOption,
  content_hash: String,
}

impl WGSLBindgen {
  fn new(options: WgslBindgenOption) -> Result<Self, WgslBindgenError> {
    let entry_points = options
      .entry_points
      .iter()
      .cloned()
      .map(SourceFilePath::new)
      .collect();

    let dependency_tree = DependencyTree::try_build(
      options.module_import_root.clone(),
      entry_points,
      options.additional_scan_dirs.clone(),
    )?;

    let content_hash = Self::get_contents_hash(&options, &dependency_tree);

    if options.emit_rerun_if_change {
      for file in Self::iter_files_to_watch(&dependency_tree) {
        println!("cargo:rerun-if-changed={}", file);
      }
    }

    Ok(Self {
      dependency_tree,
      options,
      content_hash,
    })
  }

  fn iter_files_to_watch(dep_tree: &DependencyTree) -> impl Iterator<Item = String> {
    dep_tree
      .all_files_including_dependencies()
      .into_iter()
      .map(|path| path.to_string())
  }

  fn get_contents_hash(options: &WgslBindgenOption, dep_tree: &DependencyTree) -> String {
    let mut hasher = blake3::Hasher::new();

    hasher.update(format!("{:?}", options).as_bytes());
    hasher.update(PKG_VER.as_bytes());

    for SourceFile { content, .. } in dep_tree.parsed_files() {
      hasher.update(content.as_bytes());
    }

    hasher.finalize().to_string()
  }

  fn generate_naga_module_for_entry(
    entry: SourceWithFullDependenciesResult<'_>,
  ) -> Result<WgslEntryResult, WgslBindgenError> {
    let map_err = |composer: &Composer, err: ComposerError| {
      let msg = err.emit_to_string(composer);
      WgslBindgenError::NagaModuleComposeError {
        entry: entry.source_file.file_path.to_string(),
        inner: err.inner,
        msg,
      }
    };

    let mut composer = Composer::default();
    let source = entry.source_file;

    for dependency in entry.full_dependencies.iter() {
      composer
        .add_composable_module(ComposableModuleDescriptor {
          source: &dependency.content,
          file_path: &dependency.file_path.to_string(),
          language: ShaderLanguage::Wgsl,
          as_name: dependency.module_name.as_ref().map(|name| name.to_string()),
          ..Default::default()
        })
        .map(|_| ())
        .map_err(|err| map_err(&composer, err))?;
    }

    let module = composer
      .make_naga_module(NagaModuleDescriptor {
        source: &source.content,
        file_path: &source.file_path.to_string(),
        ..Default::default()
      })
      .map_err(|err| map_err(&composer, err))?;

    Ok(WgslEntryResult {
      mod_name: source.file_path.file_prefix(),
      naga_module: module,
      source_including_deps: entry,
    })
  }

  pub fn generate_string(&self) -> Result<String, WgslBindgenError> {
    use std::fmt::Write;
    let entry_results = self
      .dependency_tree
      .get_source_files_with_full_dependencies()
      .into_iter()
      .map(Self::generate_naga_module_for_entry)
      .collect::<Result<Vec<_>, _>>()?;

    let mut text = String::new();

    if !self.options.skip_header_comments {
      writeln!(&mut text, "// File automatically generated by {PKG_NAME}^").unwrap();
      writeln!(&mut text, "//").unwrap();
      writeln!(&mut text, "// ^ {PKG_NAME} version {PKG_VER}",).unwrap();
      writeln!(&mut text, "// Changes made to this file will not be saved.").unwrap();
      writeln!(&mut text, "// SourceHash: {}", self.content_hash).unwrap();
      writeln!(&mut text).unwrap();
    }

    let output = create_rust_bindings(entry_results, &self.options)?;
    text += &output;

    Ok(text)
  }

  pub fn generate(&self) -> Result<(), WgslBindgenError> {
    let output_path = self
      .options
      .output_file
      .as_ref()
      .ok_or(WgslBindgenError::OutputFileNotSpecified)?;

    let old_content =
      std::fs::read_to_string(output_path).unwrap_or_else(|_| String::new());

    let old_hashstr_comment = old_content
      .lines()
      .find(|line| line.starts_with("// SourceHash:"))
      .unwrap_or("");

    let is_hash_changed =
      || old_hashstr_comment != format!("// SourceHash: {}", &self.content_hash);

    if self.options.skip_hash_check || is_hash_changed() {
      let content = self.generate_string()?;
      std::fs::File::create(output_path)?.write_all(content.as_bytes())?
    }

    Ok(())
  }
}
