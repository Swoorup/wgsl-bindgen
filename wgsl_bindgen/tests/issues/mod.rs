use std::fs::read_to_string;

use miette::{IntoDiagnostic, Result};
use syn::parse_str;
use wgsl_bindgen::{assert_tokens_snapshot, *};

#[test]
fn test_issue_35_short_constructor() -> Result<()> {
  WgslBindgenOptionBuilder::default()
    .workspace_root("tests/shaders/issues/issue_35")
    .add_entry_point("tests/shaders/issues/issue_35/clear.wgsl")
    .skip_hash_check(true)
    .serialization_strategy(WgslTypeSerializeStrategy::Bytemuck)
    .type_map(GlamWgslTypeMap)
    .short_constructor(2)
    .shader_source_type(WgslShaderSourceType::EmbedSource)
    .derive_serde(false)
    .emit_rerun_if_change(false)
    .skip_header_comments(true)
    .output("tests/output/issues/issue_35.actual.rs")
    .build()?
    .generate()
    .into_diagnostic()?;

  let actual = read_to_string("tests/output/issues/issue_35.actual.rs").unwrap();
  let parsed_output = parse_str(&actual).unwrap();
  assert_tokens_snapshot!(parsed_output);
  assert_rust_compilation!(parsed_output);
  Ok(())
}

#[test]
fn test_builtin_vertex_encase_issue() -> Result<()> {
  WgslBindgenOptionBuilder::default()
    .workspace_root("tests/shaders/issues")
    .add_entry_point("tests/shaders/issues/builtin_vertex_issue.wgsl")
    .skip_hash_check(true)
    .serialization_strategy(WgslTypeSerializeStrategy::Encase)
    .type_map(GlamWgslTypeMap)
    .derive_serde(false)
    .emit_rerun_if_change(false)
    .skip_header_comments(true)
    .output("tests/output/issues/builtin_vertex_encase.actual.rs")
    .build()?
    .generate()
    .into_diagnostic()?;

  let actual =
    read_to_string("tests/output/issues/builtin_vertex_encase.actual.rs").unwrap();
  let parsed_output = parse_str(&actual).unwrap();
  assert_tokens_snapshot!(parsed_output);
  // assert_rust_compilation!(parsed_output); // TODO: Fix this test
  Ok(())
}

#[test]
fn test_mixed_builtin_encase_issue() -> Result<()> {
  WgslBindgenOptionBuilder::default()
    .workspace_root("tests/shaders/issues")
    .add_entry_point("tests/shaders/issues/mixed_builtin_issue.wgsl")
    .skip_hash_check(true)
    .serialization_strategy(WgslTypeSerializeStrategy::Encase)
    .type_map(GlamWgslTypeMap)
    .derive_serde(false)
    .emit_rerun_if_change(false)
    .skip_header_comments(true)
    .output("tests/output/issues/mixed_builtin_encase.actual.rs")
    .build()?
    .generate()
    .into_diagnostic()?;

  let actual =
    read_to_string("tests/output/issues/mixed_builtin_encase.actual.rs").unwrap();
  let parsed_output = parse_str(&actual).unwrap();
  assert_tokens_snapshot!(parsed_output);
  // assert_rust_compilation!(parsed_output); // TODO: Fix this test
  Ok(())
}

#[test]
fn test_builtin_vertex_bytemuck_issue() -> Result<()> {
  WgslBindgenOptionBuilder::default()
    .workspace_root("tests/shaders/issues")
    .add_entry_point("tests/shaders/issues/builtin_vertex_bytemuck.wgsl")
    .skip_hash_check(true)
    .serialization_strategy(WgslTypeSerializeStrategy::Bytemuck)
    .type_map(GlamWgslTypeMap)
    .derive_serde(false)
    .emit_rerun_if_change(false)
    .skip_header_comments(true)
    .output("tests/output/issues/builtin_vertex_bytemuck.actual.rs")
    .build()?
    .generate()
    .into_diagnostic()?;

  let actual =
    read_to_string("tests/output/issues/builtin_vertex_bytemuck.actual.rs").unwrap();

  // Verify the struct is empty and doesn't have problematic padding bytes
  assert!(actual.contains("pub struct VertexInput {}"));

  let parsed_output = parse_str(&actual).unwrap();
  assert_tokens_snapshot!(parsed_output);
  assert_rust_compilation!(parsed_output);
  Ok(())
}

#[test]
fn test_multiple_vertex_shaders_issue() -> Result<()> {
  WgslBindgenOptionBuilder::default()
    .workspace_root("tests/shaders/issues")
    .add_entry_point("tests/shaders/issues/multiple_vertex_shaders.wgsl")
    .skip_hash_check(true)
    .serialization_strategy(WgslTypeSerializeStrategy::Bytemuck)
    .type_map(GlamWgslTypeMap)
    .derive_serde(false)
    .emit_rerun_if_change(false)
    .skip_header_comments(true)
    .output("tests/output/issues/multiple_vertex_shaders.actual.rs")
    .build()?
    .generate()
    .into_diagnostic()?;

  let actual =
    read_to_string("tests/output/issues/multiple_vertex_shaders.actual.rs").unwrap();

  // Verify correct behavior:
  // 1. Both VertexInput and InstanceInput should have VERTEX_ATTRIBUTES implementations
  assert!(actual.contains("impl VertexInput"));
  assert!(actual.contains("impl InstanceInput"));
  assert!(actual.contains("pub const VERTEX_ATTRIBUTES: [wgpu::VertexAttribute; 1]"));

  // 2. dummy_vertex_shader_entry should have VertexEntry<1> with only VertexInput
  assert!(actual.contains(
    "fn dummy_vertex_shader_entry(vertex_input: wgpu::VertexStepMode) -> VertexEntry<1>"
  ));

  // 3. dummy_instanced_vertex_shader_entry should have VertexEntry<2> with both inputs
  let has_correct_instanced_entry = actual.contains("fn dummy_instanced_vertex_shader_entry(\n    vertex_input: wgpu::VertexStepMode,\n    instance_input: wgpu::VertexStepMode,\n  ) -> VertexEntry<2>") ||
                                   actual.contains("fn dummy_instanced_vertex_shader_entry(vertex_input: wgpu::VertexStepMode, instance_input: wgpu::VertexStepMode) -> VertexEntry<2>");
  assert!(has_correct_instanced_entry);

  let parsed_output = parse_str(&actual).unwrap();
  assert_tokens_snapshot!(parsed_output);
  assert_rust_compilation!(parsed_output);
  Ok(())
}

#[test]
fn test_vec3a_padding_overflow_issue() -> Result<()> {
  WgslBindgenOptionBuilder::default()
    .workspace_root("tests/shaders/issues")
    .add_entry_point("tests/shaders/issues/vec3a_padding_issue.wgsl")
    .skip_hash_check(true)
    .serialization_strategy(WgslTypeSerializeStrategy::Bytemuck)
    .type_map(GlamWgslTypeMap)
    .derive_serde(false)
    .emit_rerun_if_change(false)
    .skip_header_comments(true)
    .output("tests/output/issues/vec3a_padding_overflow.actual.rs")
    .build()?
    .generate()
    .into_diagnostic()?;

  let actual =
    read_to_string("tests/output/issues/vec3a_padding_overflow.actual.rs").unwrap();
  let parsed_output = parse_str(&actual).unwrap();
  assert_tokens_snapshot!(parsed_output);
  assert_rust_compilation!(parsed_output);
  Ok(())
}

#[test]
fn test_duplicate_import_vertexinput_issue() -> Result<()> {
  // This test reproduces the issue where importing the same VertexInput struct
  // into multiple shader files causes "duplicate content found" error
  println!("Testing duplicate import issue...");

  let actual = WgslBindgenOptionBuilder::default()
    .workspace_root("tests/shaders/issues/duplicate_import")
    .add_entry_point("tests/shaders/issues/duplicate_import/shader1.wgsl")
    .add_entry_point("tests/shaders/issues/duplicate_import/shader2.wgsl")
    .skip_hash_check(true)
    .serialization_strategy(WgslTypeSerializeStrategy::Bytemuck)
    .type_map(GlamWgslTypeMap)
    .derive_serde(false)
    .emit_rerun_if_change(false)
    .skip_header_comments(true)
    .output("tests/output/issues/duplicate_import.actual.rs")
    .build()
    .unwrap()
    .generate_string()
    .into_diagnostic()
    .unwrap();

  // For now, just check that we got some output - we'll refine assertions once we see what's generated
  assert!(!actual.is_empty(), "Generated code should not be empty");

  // Try to parse and compile
  let parsed_output = parse_str(&actual).unwrap();
  assert_rust_compilation!(parsed_output);

  Ok(())
}
