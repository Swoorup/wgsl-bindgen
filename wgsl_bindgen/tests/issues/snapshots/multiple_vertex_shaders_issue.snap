---
source: wgsl_bindgen/tests/issues/mod.rs
---
#![allow(unused, non_snake_case, non_camel_case_types, non_upper_case_globals)]
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum ShaderEntry {
  MultipleVertexShaders,
}
impl ShaderEntry {
  pub fn create_pipeline_layout(&self, device: &wgpu::Device) -> wgpu::PipelineLayout {
    match self {
      Self::MultipleVertexShaders => {
        multiple_vertex_shaders::create_pipeline_layout(device)
      }
    }
  }
  pub fn create_shader_module_embed_source(
    &self,
    device: &wgpu::Device,
  ) -> wgpu::ShaderModule {
    match self {
      Self::MultipleVertexShaders => {
        multiple_vertex_shaders::create_shader_module_embed_source(device)
      }
    }
  }
}
mod _root {
  pub use super::*;
}
pub mod layout_asserts {
  use super::{_root, _root::*};
  const WGSL_BASE_TYPE_ASSERTS: () = {
    assert!(std::mem::size_of::<glam::IVec2>() == 8);
    assert!(std::mem::align_of::<glam::IVec2>() == 4);
    assert!(std::mem::size_of::<glam::IVec3>() == 12);
    assert!(std::mem::align_of::<glam::IVec3>() == 4);
    assert!(std::mem::size_of::<glam::IVec4>() == 16);
    assert!(std::mem::align_of::<glam::IVec4>() == 4);
    assert!(std::mem::size_of::<glam::UVec2>() == 8);
    assert!(std::mem::align_of::<glam::UVec2>() == 4);
    assert!(std::mem::size_of::<glam::UVec3>() == 12);
    assert!(std::mem::align_of::<glam::UVec3>() == 4);
    assert!(std::mem::size_of::<glam::UVec4>() == 16);
    assert!(std::mem::align_of::<glam::UVec4>() == 4);
    assert!(std::mem::size_of::<glam::Vec2>() == 8);
    assert!(std::mem::align_of::<glam::Vec2>() == 4);
    assert!(std::mem::size_of::<glam::Vec3>() == 12);
    assert!(std::mem::align_of::<glam::Vec3>() == 4);
    assert!(std::mem::size_of::<glam::Vec4>() == 16);
    assert!(std::mem::align_of::<glam::Vec4>() == 16);
    assert!(std::mem::size_of::<glam::Mat2>() == 16);
    assert!(std::mem::align_of::<glam::Mat2>() == 16);
    assert!(std::mem::size_of::<glam::Mat3A>() == 48);
    assert!(std::mem::align_of::<glam::Mat3A>() == 16);
    assert!(std::mem::size_of::<glam::Mat4>() == 64);
    assert!(std::mem::align_of::<glam::Mat4>() == 16);
  };
}
pub mod multiple_vertex_shaders {
  use super::{_root, _root::*};
  #[repr(C)]
  #[derive(Debug, PartialEq, Clone, Copy)]
  pub struct VertexInput {
    pub position: glam::Vec3,
  }
  impl VertexInput {
    pub const fn new(position: glam::Vec3) -> Self {
      Self { position }
    }
  }
  impl VertexInput {
    pub const VERTEX_ATTRIBUTES: [wgpu::VertexAttribute; 1] = [wgpu::VertexAttribute {
      format: wgpu::VertexFormat::Float32x3,
      offset: std::mem::offset_of!(Self, position) as u64,
      shader_location: 0,
    }];
    pub const fn vertex_buffer_layout(
      step_mode: wgpu::VertexStepMode,
    ) -> wgpu::VertexBufferLayout<'static> {
      wgpu::VertexBufferLayout {
        array_stride: std::mem::size_of::<Self>() as u64,
        step_mode,
        attributes: &Self::VERTEX_ATTRIBUTES,
      }
    }
  }
  #[repr(C)]
  #[derive(Debug, PartialEq, Clone, Copy)]
  pub struct InstanceInput {
    pub position: glam::Vec3,
  }
  impl InstanceInput {
    pub const fn new(position: glam::Vec3) -> Self {
      Self { position }
    }
  }
  impl InstanceInput {
    pub const VERTEX_ATTRIBUTES: [wgpu::VertexAttribute; 1] = [wgpu::VertexAttribute {
      format: wgpu::VertexFormat::Float32x3,
      offset: std::mem::offset_of!(Self, position) as u64,
      shader_location: 1,
    }];
    pub const fn vertex_buffer_layout(
      step_mode: wgpu::VertexStepMode,
    ) -> wgpu::VertexBufferLayout<'static> {
      wgpu::VertexBufferLayout {
        array_stride: std::mem::size_of::<Self>() as u64,
        step_mode,
        attributes: &Self::VERTEX_ATTRIBUTES,
      }
    }
  }
  pub const ENTRY_DUMMY_VERTEX_SHADER: &str = "dummy_vertex_shader";
  pub const ENTRY_DUMMY_INSTANCED_VERTEX_SHADER: &str = "dummy_instanced_vertex_shader";
  #[derive(Debug)]
  pub struct VertexEntry<const N: usize> {
    pub entry_point: &'static str,
    pub buffers: [wgpu::VertexBufferLayout<'static>; N],
    pub constants: Vec<(&'static str, f64)>,
  }
  pub fn vertex_state<'a, const N: usize>(
    module: &'a wgpu::ShaderModule,
    entry: &'a VertexEntry<N>,
  ) -> wgpu::VertexState<'a> {
    wgpu::VertexState {
      module,
      entry_point: Some(entry.entry_point),
      buffers: &entry.buffers,
      compilation_options: wgpu::PipelineCompilationOptions {
        constants: &entry.constants,
        ..Default::default()
      },
    }
  }
  pub fn dummy_vertex_shader_entry(vertex_input: wgpu::VertexStepMode) -> VertexEntry<1> {
    VertexEntry {
      entry_point: ENTRY_DUMMY_VERTEX_SHADER,
      buffers: [VertexInput::vertex_buffer_layout(vertex_input)],
      constants: Default::default(),
    }
  }
  pub fn dummy_instanced_vertex_shader_entry(
    vertex_input: wgpu::VertexStepMode,
    instance_input: wgpu::VertexStepMode,
  ) -> VertexEntry<2> {
    VertexEntry {
      entry_point: ENTRY_DUMMY_INSTANCED_VERTEX_SHADER,
      buffers: [
        VertexInput::vertex_buffer_layout(vertex_input),
        InstanceInput::vertex_buffer_layout(instance_input),
      ],
      constants: Default::default(),
    }
  }
  #[derive(Debug)]
  pub struct WgpuPipelineLayout;
  impl WgpuPipelineLayout {
    pub fn bind_group_layout_entries(
      entries: [wgpu::BindGroupLayout; 0],
    ) -> [wgpu::BindGroupLayout; 0] {
      entries
    }
  }
  pub fn create_pipeline_layout(device: &wgpu::Device) -> wgpu::PipelineLayout {
    device.create_pipeline_layout(&wgpu::PipelineLayoutDescriptor {
      label: Some("MultipleVertexShaders::PipelineLayout"),
      bind_group_layouts: &[],
      push_constant_ranges: &[],
    })
  }
  pub fn create_shader_module_embed_source(device: &wgpu::Device) -> wgpu::ShaderModule {
    let source = std::borrow::Cow::Borrowed(SHADER_STRING);
    device.create_shader_module(wgpu::ShaderModuleDescriptor {
      label: Some("multiple_vertex_shaders.wgsl"),
      source: wgpu::ShaderSource::Wgsl(source),
    })
  }
  pub const SHADER_STRING: &str = r#"
struct VertexInput {
    @location(0) position: vec3<f32>,
}

struct InstanceInput {
    @location(1) position: vec3<f32>,
}

@vertex 
fn dummy_vertex_shader(vert_in: VertexInput) -> @builtin(position) vec4<f32> {
    return vec4<f32>(vert_in.position, 1f);
}

@vertex 
fn dummy_instanced_vertex_shader(vert_in_1: VertexInput, instance_in: InstanceInput) -> @builtin(position) vec4<f32> {
    return vec4<f32>((vert_in_1.position + instance_in.position), 1f);
}
"#;
}
pub mod bytemuck_impls {
  use super::{_root, _root::*};
  unsafe impl bytemuck::Zeroable for multiple_vertex_shaders::VertexInput {}
  unsafe impl bytemuck::Pod for multiple_vertex_shaders::VertexInput {}
  unsafe impl bytemuck::Zeroable for multiple_vertex_shaders::InstanceInput {}
  unsafe impl bytemuck::Pod for multiple_vertex_shaders::InstanceInput {}
}
